vendor/github.com/prometheus/prometheus/tsdb/wlog/wlog.go

```diff
@@ -1046,3 +1046,32 @@ func (r *segmentBufReader) Read(b []byte) (n int, err error) {
 func (w *WL) Size() (int64, error) {
 	return fileutil.DirSize(w.Dir())
 }
+
+func (w *WL) FsyncSegmentsUntilCurrent() error {
+	var err error
+	w.mtx.Lock()
+	if w.closed {
+		w.mtx.Unlock()
+		return errors.New("unable to fsync segments: write log is closed")
+	}
+	// fsync current segment within mutex to avoid race conditions where the segment could be updated or closed
+	if w.segment != nil {
+		err = w.fsync(w.segment)
+	}
+	w.mtx.Unlock()
+
+	if err != nil {
+		return err
+	}
+
+	done := make(chan struct{})
+	// all previous segments before w.segment should either have been fsynced and closed or still in the actorc queue
```

Comment:
For consistency, you could do the `fsync` inside the actor function.
I can't come up with any concrete reason why this would be better; it just seems nicer that all fsyncs would happen in order.